File: src/main/java/cn/solarmoon/idyllic_food_diary/element/matter/cookware/kettle/KettleBlock.java
package cn.solarmoon.idyllic_food_diary.element.matter.cookware.kettle;

import cn.solarmoon.idyllic_food_diary.element.matter.cookware.CookwareBlock;
import cn.solarmoon.idyllic_food_diary.registry.common.IMParticles;
import cn.solarmoon.idyllic_food_diary.registry.common.IMBlockEntities;
import cn.solarmoon.idyllic_food_diary.registry.common.IMSounds;
import cn.solarmoon.solarmoon_core.api.blockstate_access.IHorizontalFacingBlock;
import cn.solarmoon.solarmoon_core.api.tile.fluid.FluidHandlerUtil;
import cn.solarmoon.solarmoon_core.api.tile.inventory.ItemHandlerUtil;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.SoundType;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.material.MapColor;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.Vec3;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.phys.shapes.VoxelShape;
import org.jetbrains.annotations.NotNull;

import java.util.Random;

/**
 * 基本的烧水壶抽象类
 * 应用tankBlockEntity，是一个液体容器（但是不使其具有原有的液体交互功能，而是替换为倒水功能）
 * 具有烧水、倒水功能
 */
public class KettleBlock extends CookwareBlock {

    public KettleBlock() {
        super(Properties.of()
                .sound(SoundType.LANTERN)
                .strength(2f, 6.0F)
                .mapColor(MapColor.METAL)
                .forceSolidOn()
        );
    }

    @Override
    public InteractionResult originUse(BlockState state, Level level, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hitResult) {
        BlockEntity blockEntity = level.getBlockEntity(pos);
        KettleBlockEntity kettle = (KettleBlockEntity) blockEntity;
        if(kettle == null) return InteractionResult.PASS;

        //液体交互
        if (FluidHandlerUtil.loadFluid(kettle.getTank(), player, hand, false)) {
            level.playSound(null, pos, SoundEvents.BUCKET_FILL, SoundSource.PLAYERS);
            return InteractionResult.SUCCESS;
        }
        if (hand == InteractionHand.MAIN_HAND && ItemHandlerUtil.storage(kettle.getInventory(), player, hand, 1, 1)) {
            level.playSound(null, pos, SoundEvents.LANTERN_HIT, SoundSource.PLAYERS);
            return InteractionResult.SUCCESS;
        }

        return InteractionResult.PASS;
    }

    @Override
    public void tick(Level level, BlockPos pos, BlockState state, BlockEntity blockEntity) {
        super.tick(level, pos, state, blockEntity);
        KettleBlockEntity kettle = (KettleBlockEntity) blockEntity;

        if (!kettle.tryBrewTea()) kettle.tryBoilWater();

        //里面是hot类型的液体就冒热气
        makeBoilParticle(level, pos, state, blockEntity);
    }

    /**
     * 烧水粒子
     */
    protected static void makeBoilParticle(Level level, BlockPos pos, BlockState state, BlockEntity blockEntity) {
        Direction face = state.getValue(KettleBlock.FACING);
        float delta1 = -face.getStepZ() * 0.5f;
        float delta2 = face.getStepX() * 0.5f;
        Vec3 spoutPos = pos.getCenter().add(delta1, 0.07, delta2);

        if (blockEntity instanceof KettleBlockEntity kettle) {
            Random random = new Random();
            if (kettle.isInBoil()) {
                if (random.nextFloat() < 0.8) {
                    level.addParticle(IMParticles.CRASHLESS_CLOUD.get(), spoutPos.x, spoutPos.y, spoutPos.z, 0, 0.1, 0);
                }
                if (random.nextFloat() < 0.1) {
                    level.playSound(null, pos, IMSounds.BOILING_WATER.get(), SoundSource.BLOCKS, 1F, 1F);
                }
            }
            else if (kettle.isBoiling()) {
                if (random.nextFloat() < 0.1)
                    level.addParticle(IMParticles.CRASHLESS_CLOUD.get(), spoutPos.x, spoutPos.y, spoutPos.z, 0, 0.01, 0);
            }

        }
    }

    protected static final VoxelShape SHAPE = Block.box(4.0D, 0.0D, 4.0D, 12.0D, 7.0D, 12.0D);
    @Override
    public @NotNull VoxelShape getOriginShape(@NotNull BlockState state, @NotNull BlockGetter level, @NotNull BlockPos pos, @NotNull CollisionContext context) {
        return SHAPE;
    }

    @Override
    public BlockEntityType<?> getBlockEntityType() {
        return IMBlockEntities.KETTLE.get();
    }

}

File: src/main/kotlin/cn/solarmoon/idyllic_food_diary/element/matter/cookware/kettle/KettleBlock.kt
package cn.solarmoon.idyllic_food_diary.element.matter.cookware.kettle

import cn.solarmoon.idyllic_food_diary.element.matter.cookware.CookwareBlock
import cn.solarmoon.idyllic_food_diary.registry.common.IFDBlockEntities
import cn.solarmoon.idyllic_food_diary.registry.common.IFDParticles
import cn.solarmoon.idyllic_food_diary.registry.common.IFDSounds
import cn.solarmoon.spark_core.api.blockstate.IHorizontalFacingState
import cn.solarmoon.spark_core.api.cap.item.ItemStackHandlerHelper
import net.minecraft.core.BlockPos
import net.minecraft.sounds.SoundEvents
import net.minecraft.sounds.SoundSource
import net.minecraft.world.InteractionHand
import net.minecraft.world.ItemInteractionResult
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.BlockGetter
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.SoundType
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.phys.BlockHitResult
import net.minecraft.world.phys.shapes.CollisionContext
import net.minecraft.world.phys.shapes.VoxelShape
import net.neoforged.neoforge.fluids.FluidUtil
import java.util.Random

class KettleBlock: CookwareBlock(Properties.of().sound(SoundType.LANTERN).strength(2f).forceSolidOn()) {

    override fun useItemOnThis(
        heldItem: ItemStack,
        state: BlockState,
        level: Level,
        pos: BlockPos,
        player: Player,
        hand: InteractionHand,
        hitResult: BlockHitResult
    ): ItemInteractionResult {
        val kettle = level.getBlockEntity(pos) as KettleBlockEntity
        if (FluidUtil.interactWithFluidHandler(player, hand, level, pos, hitResult.direction)) return ItemInteractionResult.sidedSuccess(level.isClientSide)
        if (ItemStackHandlerHelper.storage(kettle.inventory, player, hand, 1, 1)) {
            level.playSound(null, pos, getSoundType(state, level, pos, null).hitSound, SoundSource.PLAYERS)
            return ItemInteractionResult.sidedSuccess(level.isClientSide)
        }
        return ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION
    }

    override fun tick(level: Level, pos: BlockPos, state: BlockState, blockEntity: BlockEntity) {
        super.tick(level, pos, state, blockEntity)
        val kettle = blockEntity as KettleBlockEntity

        if (!kettle.boil.tryWork()) kettle.brew.tryWork()
        makeBoilParticle(level, pos, state, kettle)
    }

    fun makeBoilParticle(level: Level, pos: BlockPos, state: BlockState, kettle: KettleBlockEntity) {
        val face = state.getValue(IHorizontalFacingState.FACING);
        val delta1 = -face.stepZ * 0.5
        val delta2 = face.stepX * 0.5
        val spoutPos = pos.center.add(delta1, 0.07, delta2);
        val random = Random()
        if (kettle.boil.isInBoil()) {
            if (random.nextFloat() < 0.8) {
                level.addAlwaysVisibleParticle(IFDParticles.CRASHLESS_CLOUD.get(), spoutPos.x, spoutPos.y, spoutPos.z, 0.0, 0.1, 0.0);
            }
            if (random.nextFloat() < 0.1) {
                level.playSound(null, pos, IFDSounds.BOILING_WATER.get(), SoundSource.BLOCKS)
            }
        }
        else if (kettle.boil.isBoiling()) {
            if (random.nextFloat() < 0.1)
                level.addAlwaysVisibleParticle(IFDParticles.CRASHLESS_CLOUD.get(), spoutPos.x, spoutPos.y, spoutPos.z, 0.0, 0.01, 0.0);
        }
    }

    override fun getShapeThis(
        state: BlockState,
        level: BlockGetter,
        pos: BlockPos,
        context: CollisionContext
    ): VoxelShape {
        return box(4.0, 0.0, 4.0, 12.0, 7.0, 12.0)
    }

    override fun getBlockEntityType(): BlockEntityType<*> {
        return IFDBlockEntities.KETTLE.get()
    }

}